src/Build/Evaluation/Expander.cs

```diff
@@ -556,24 +556,8 @@ private static int ScanForClosingParenthesis(string expression, int index, out b
         /// </summary>
         private static int ScanForClosingQuote(char quoteChar, string expression, int index)
         {
-            unsafe
-            {
-                fixed (char* pchar = expression)
-                {
-                    // Scan for our closing quoteChar
-                    while (index < expression.Length)
-                    {
-                        if (pchar[index] == quoteChar)
```

Comment:
I was under the impression that this sort of thing was supposed to be fast. What makes IndexOf faster?

Reply:
What we had here was a naïve implementation of `IndexOf`. This is the real thing:

https://github.com/dotnet/runtime/blob/1bff0cce7ba7de077c09e331d24e3b27c3779888/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs#L213-L493

I bet it's faster because of 1) HW acceleration and 2) lack of pinning. Making character enumeration faster by pinning the string and accessing the `char *` with unsafe code might have been the case in very early versions of .NET but it is counter-productive now as it prevents the GC from moving the object.

It's OK to not use advanced vector code when writing custom specialized string search functions. But if the function is literally `IndexOf`, there is absolutely no reason why it shouldn't just call `IndexOf`.

Reply:
I didn't realize the GC moves strings around—presumably to reduce fragmentation? When I built my simplistic GC for a class, it tried to combine consecutive chunks of free memory to make larger chunks, but it certainly wouldn't move anything. I'm curious how much of an improvement that made, since less frequent GCs would presumably be partially balanced by the cost of moving things, though that's the sort of task computers excel at.

Reply:
The .NET GC indeed moves objects. When promoting to a higher generation for example, as Gen0/1 and Gen2 are always in different segments.
https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals#what-happens-during-a-garbage-collection